### ISSUE TITLE
WLink "memory could not be read" exception

### ISSUE THREAD

**Issue by Pokeman2003** [open-watcom/open-watcom-v2#1360]
I'm using the [December 2nd 32-bit Windows build](https://github.com/open-watcom/open-watcom-v2/releases/tag/2024-12-02-Build).
I have a basic linker script that looks like this. By all accounts, I think this should link. It doesn't look too dissimilar to the examples provided, other than the absurd list of libraries I included to stop a different error.
```
system nt
option map
name output
file W:\Project\CompileFolder\Backend\File\Ingress.obj
file W:\Project\CompileFolder\Backend\Localization\LocalEN.obj
file W:\Project\CompileFolder\Backend\Localization\LocalFallback.obj
file W:\Project\CompileFolder\Windows\Initialize.obj
library W:\WATCOM\CC++\lib386\cplx3r.lib
library W:\WATCOM\CC++\lib386\cplx3s.lib
# Skipped the majority of these, since it's literally just a dump of every *.lib file from lib386 and lib386\nt.
# I wish I knew what libraries were actually necessary.
library W:\WATCOM\CC++\lib386\nt\wsock32.lib
library W:\WATCOM\CC++\lib386\nt\wtsapi32.lib
```
Instead of linking, my linker has a very, very angry output instead. On top of the [non-sense warning](https://open-watcom.github.io/open-watcom-v2-wikidocs/lguide.pdf#page=17), it also spits out an exception.
```
Warning! W1107: file W:\Project\CompileFolder\link: line(1): undefined system name: nt
Warning! W1014: stack segment not found
The instruction at 0x69c2200e referenced memory at 0x00000008.
The memory could not be read.
Exception fielded by 0x00403000
EAX=0x00000000 EBX=0x0019fcd1 ECX=0x00000000 EDX=0x0019fcc0
ESI=0x0019fd30 EDI=0x0019fcd0 EBP=0x0019fd18 ESP=0x0019fc80
EIP=0x69c2200e EFL=0x00010246 CS =0x00000023 SS =0x0000002b
DS =0x0000002b ES =0x0000002b FS =0x00000053 GS =0x0000002b
Stack dump (SS:ESP)
0x00000000 0x0000ffff 0x00000000 0x69c2d193 0x00000000 0x004757b9
0x00040000 0x0019fd60 0x0019fcc0 0x69c21007 0x0019fcec 0x0019fd00
0x0019fd18 0x0019fcec 0x0019fd18 0x69c22639 0x00000008 0x00000004
0x000000b8 0x00b80007 0x0019fd18 0x0019fcec 0x00472510 0x00472510
0x00000000 0x00000006 0x69c22796 0x000015ff 0x69c50000 0x0019fd8c
0x00000002 0x0019fd60 0x00000008 0x0019fd60 0x0019fd28 0x69c5445c
0x00000008 0x69c229b2 0x0019fcee 0x00000000 0x000000b8 0x00000001
0x00000000 0x0000ffff 0x06040202 0x00000400 0x00000000 0x0000ffff
0x06040202 0x00000008 0x0019fd68 0x0019fe48 0x004020ca 0x0019fe20
0x0047e184 0x69c22af6 0x00590311 0x69c38502 0x69c544f0 0x69c54406
0x06040202 0x0047e184 0x0000006c 0x69c22a40 0x69c2bfe3 0x001f11b0
0x69c28cc0 0x004741d1 0x00000000 0x69c28c8e 0x69c33b80 0x69c28c62
```
[Here is the linking environment,](https://github.com/user-attachments/files/18203527/CompileFolder.zip) and [here is my compilation environment](https://github.com/user-attachments/files/18203508/Source.zip) if you so desire.

**Comment by Pokeman2003** [comment]
I think I know what happened, actually. Let me do a bit more testing.

**Comment by Pokeman2003** [comment]
What's the difference between the R and S libraries? I suspect mixing them is the problem.

**Comment by jmalak** [comment]
Thanks for your bug report.
The crash is not OK it will be fixed.

Anyway from the first message
Warning! W1107: file W:\Project\CompileFolder\link: line(1): undefined system name: nt
is clear that you have not properly configured Open Watcom and wlink has not defined any system due to missing definition file.
Please configure Open Watcom properly.
- WATCOM environment should be defined and point to the root of your OW installation
- PATH must contains OW binary subdirectory which contains OW binaries for your host Operating system
- INCLUDE must contains OW header subdirectory which contains OW header files for your target Operating system

**Comment by Pokeman2003** [comment]
I am desperately trying to avoid environment variables for portability reasons. I've managed to pass the correct header files using a -i inclusion, so I've at least gotten that far. If you know, why are the others necessary?

**Comment by Pokeman2003** [comment]
Git forces me to use an absolute path, so I began using virtual drives to organize my projects because it makes switching between my laptop and desktop easier. I also prefer to keep the tools on the virtual drive, despite redundancy, in case I want to program on a machine that doesn't have these tools.

Something to note is that Windows and DOS are both operating systems that are primarily meant to handle portable installations. Yeah, there's plenty of programs that ignore this convention, but it's like how alt+enter is supposed to be the fully screen keybind by Windows' own convention, but few programs actually use this keybind for that.

Anyways all of this to say is that I'm running one last test because I think someone may have conflated an environment variable and a regular variable in the manual back when it was first written. There is a huge difference between the two(you couldn't even set the former in Batch before Windows 7,) but the linker manual has at least a hint suggesting they may be taking from the regular variables instead. If my test fails, no harm no foul I'll just redo the installation.

**Comment by jmalak** [comment]
We are not depending on any OS specific because Open Watcom is by nature cross-compiler.
By example you install it on DOS and run only setup batch and you can compile to any target on DOS.
If you install it on Linux same, you run setup script and you can compile to any target.
Nothing specific per host OS.
It is absolutely portable to any OS, only 3 environment variables need to setup.

**Comment by Pokeman2003** [comment]
That's not what portable means. Portable means that the installation is self-contained and can have its directory changed at will without affecting operations of the program.

**Comment by jmalak** [comment]
You can copy OW tree anywhere and run batch file, it is not portable?

**Comment by Pokeman2003** [comment]
Last post before I go test.
Environment variables are GLOBAL variables that are available at all times, regardless of circumstances in a Batch script. As far as I'm aware, _setenv_ has continued to be the only way to set them on Windows NT machines with batch. These are very different and can even be accessed outside of a batch call for the reason of "they are literally global to the machine/user." This is VERY MUCH not portable because it is a FIXED change, like adding a registry key.
There are REGULAR variables that are set with just a plain _set_ command. They can be expanded by batch, and while this is the first time I've ever heard of a program being able to read them outside of the built-in shell executables, it wouldn't surprise me if there was a means to do so. This is what I'm going to go test.

**Comment by jmalak** [comment]
Sorry you are wrong, environment setup can be various, temporary or user specific or system.
Open Watcom don't limit you anyway, you can use various method it depend on your requirements.
On my development system I have various OW installations and before any work I run simple appropriate setup script for 3 environment variables and nothing more after that I works with appropriate OW version without anything specific. It ensures that any make file is working properly on the host system and for target system.

**Comment by Pokeman2003** [comment]
Yeah you're right, I'm wrong. I was wondering why setenv wasn't working. Oops.

**Comment by Pokeman2003** [comment]
Yeah okay, well that solves that bit and simplifies the script down immensely. An impressively overengineered solution, defeated by a simple setlocal to make sure I don't make a mess out of my path variables. Jesus Christ, what is wrong with me...

**Comment by Pokeman2003** [comment]
Honestly I'm not even sure if you can replicate my crash without the complete misunderstanding of the very operating system you're using.
(Yes, adding the environment variables fixed the crash.)

**Comment by jmalak** [comment]
No problem, now you understand how it should be and don't create your own problems.
You need not solve such issue anymore.
You need only create portable make files and if you will move to another system only change these environment variables and all should going OK.
